Modules/_decimal/_decimal.c

```diff
@@ -3672,34 +3678,46 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
-#if PYLONG_BITS_IN_DIGIT == 30
-    n = mpd_qexport_u32(&ob_digit, 0, PyLong_BASE, x, &status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    n = mpd_qexport_u16(&ob_digit, 0, PyLong_BASE, x, &status);
-#else
-    #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
 
-    if (n == SIZE_MAX) {
-        PyErr_NoMemory();
+    int64_t val = mpd_qget_i64(x, &status);
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+    assert(!mpd_iszero(x));
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+    const uint32_t base = (uint32_t)1 << layout->bits_per_digit;
```

Comment:
To be portable, I think you can use `UINT32_C(1) << layout->bits_per_digit` instead.

Reply:
Hmm, why do you think so?  Type cast should work fine and I see not too much cases for using `UINT**_C` macros in the CPython codebase.

Reply:
Oh, I thought we use it in general for this. But I guess it's fine if this is not the usual way.